exonum-java-binding/core/rust/integration_tests/tests/runtime.rs

```diff
@@ -0,0 +1,667 @@
+    let (runtime, _instance_id, config, _path) = create_runtime_with_valid_test_config();
+
+    let mut test_kit = TestKitBuilder::validator()
+        .with_additional_runtime(runtime)
+        .with_instances(vec![config])
+        .create();
+
+    let mut value: Option<i32> = None;
+
+    for _ in 0..10 {
+        test_kit.create_block();
+
+        let snapshot = test_kit.snapshot();
+        let map = create_before_commit_test_map(&*snapshot, fake_service::TEST_SERVICE_NAME);
+        let key = hash(fake_service::BEFORE_COMMIT_ENTRY_KEY.as_ref());
+        let cur_value = map
+            .get(&key)
+            .expect("Failed to find the entry created in beforeCommit")
+            .parse::<i32>()
+            .expect("Failed to parse the entry value created in beforeCommit");
+
+        if let Some(prev_value) = value {
+            assert_eq!(prev_value + 1, cur_value);
+        }
+
+        value = Some(cur_value);
+    }
+
+    test_kit.stop();
+}
+
+#[test]
+fn state_hashes() {
+    let service_hash = &[
+        Hash::from_hex("8c1ea14c7893acabde2aa95031fae57abb91516ddb78b0f6622afa0d8cb1b5c2").unwrap(),
+        Hash::from_hex("7324b5c72b51bb5d4c180f1109cfd347b60473882145841c39f3e584576296f9").unwrap(),
+    ];
+
+    let (runtime, _instance_id, config, _path) = create_runtime_with_valid_test_config();
+    let runtime_copy = runtime.clone();
+
+    let mut test_kit = TestKitBuilder::validator()
+        .with_additional_runtime(runtime)
+        .with_instances(vec![config])
+        .create();
+
+    for _ in 0..5 {
+        test_kit.create_block();
+
+        let snapshot = test_kit.snapshot();
+        let aggregator = runtime_copy.state_hashes(&snapshot);
+
+        assert_eq!(aggregator.instances.len(), 1);
+        let (instance, hashes) = aggregator
+            .instances
+            .get(0)
+            .expect("Failed to find state hash pair for test service");
+        assert_eq!(*instance, VALID_INSTANCE_ID);
+
+        assert_eq!(hashes.len(), 2);
+        let hash = hashes
+            .get(0)
+            .expect("Failed to find state_hash[0] for test-service");
+        assert_eq!(*hash, service_hash[0]);
+        let hash = hashes
+            .get(1)
+            .expect("Failed to find state_hash[1] for test-service");
+        assert_eq!(*hash, service_hash[1]);
+    }
+
+    test_kit.stop();
+}
+
+fn create_artifact<C>(
+    executor: &Executor,
+    artifact_name: &str,
+    artifact_version: &str,
+    build_fn: C,
+) -> (ArtifactId, DeployArguments, TempPath)
+where
+    C: FnOnce(&Executor, &str, &str) -> TempPath,
+{
+    let artifact = ArtifactId::new(
+        JavaRuntimeProxy::RUNTIME_ID,
+        format!("{}:{}", artifact_name, artifact_version),
+    )
+    .unwrap();
+    let path = build_fn(executor, &artifact.to_string(), artifact_version);
+    let deploy_args = DeployArguments {
+        artifact_filename: path.to_str().unwrap().to_string(),
+    };
+
+    (artifact, deploy_args, path)
```

Comment:
I am not sure `path` is used downstream. It is included in the deploy arguments — is there code using it?